### PR TITLE
feat(GC): GC remaining cols

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes https://github.com/nearprotocol/nearcore/issues/2860.

- Added ColOutcomesByBlockHash to allow getting all Outcome Ids for Block Hash
- Clearing `ColLastBlockWithNewChunk` in case when current GC Block is the last Block with new chunk
- Clearing `ColEpochInfo` and `ColEpochStart` is the Block is the last one in the Epoch - it doesn't clear data in EpochManager cache!
- db_version increased
- LastComponentNonce -> ColLastComponentNonce
- Added checks from https://github.com/nearprotocol/nearcore/issues/2893
- Added checks from https://github.com/nearprotocol/nearcore/issues/2861
- ColGCCount rewritten issue is fixed
- ColStateParts explicit clearing in GC
- Printing DBCol counters with Strum `from` impl

### CODE REVIEW

File: chain/chain/src/store.rs [hunk 1]
```diff
@@ -2368,9 +2422,31 @@ impl<'a> ChainStoreUpdate<'a> {
...
+                    hash_set.insert(outcome.id().clone());
+                    entry.insert(hash_set);
+                }
+            };
             store_update.set_ser(ColTransactionResult, hash.as_ref(), outcome)?;
         }
+        for (block_hash, hash_set) in block_hash_to_outcomes {
+            store_update
+                .set_ser(ColOutcomesByBlockHash, block_hash.as_ref(), &hash_set)
+                .map_err::<Error, _>(|e| e.into())?;
```
Comment (frol) [inline chain/chain/src/store.rs:2448]: `?` should call `.into()` automatically, so this `.map_err` seems to be redundant, doesn't it?

File: chain/chain/src/store.rs [hunk 2]
```diff
@@ -2929,8 +3006,10 @@ mod tests {
             DBCol::ColNextBlockWithNewChunk,
             DBCol::ColChunkPerHeightShard,
             DBCol::ColBlockRefCount,
+            DBCol::ColOutcomesByBlockHash,
         ];
         for i in DBCol::iter() {
+            println!("{:?}", i);
```
Comment (frol) [inline chain/chain/src/store.rs:3012]: Remove debug output

File: core/store/src/lib.rs [hunk 3]
```diff
@@ -270,6 +273,43 @@ pub fn create_store(path: &str) -> Arc<Store> {
...
 
+fn get_all_block_hash_to_outcomes(store: &Store, block_hash: &CryptoHash) -> HashSet<CryptoHash> {
+    match store.get_ser(ColOutcomesByBlockHash, block_hash.as_ref()) {
+        Ok(Some(hash_set)) => hash_set,
+        Ok(None) => HashSet::new(),
+        Err(e) => panic!("Can't read DB, {:?}", e),
+    }
+}
+
+pub fn fill_col_outcomes_by_hash(store: &Store) {
```
Comment (frol) [inline core/store/src/lib.rs:284]: Should we create a separate file for the migration utils? It seems that this method is only useful once, when we want to migrate the DB.

File: core/store/src/lib.rs [hunk 4]
```diff
@@ -270,6 +273,43 @@ pub fn create_store(path: &str) -> Arc<Store> {
...
+        })
+        .collect();
+    let mut block_hash_to_outcomes: HashMap<CryptoHash, HashSet<CryptoHash>> = HashMap::new();
+    for outcome in outcomes {
+        match block_hash_to_outcomes.entry(outcome.block_hash) {
+            Entry::Occupied(mut entry) => {
+                entry.get_mut().insert(outcome.id().clone());
+            }
+            Entry::Vacant(entry) => {
+                let mut hash_set = get_all_block_hash_to_outcomes(store, &outcome.block_hash);
```
Comment (frol) [inline core/store/src/lib.rs:300]: Can we reuse the relevant method from ChainStoreAccess?

File: chain/chain/src/store.rs [hunk 5]
```diff
@@ -229,6 +224,11 @@ pub trait ChainStoreAccess {
         &mut self,
         hash: &CryptoHash,
     ) -> Result<&ExecutionOutcomeWithIdAndProof, Error>;
+    /// Returns a HashSet of Outcome ids for current Block Hash
+    fn get_all_block_hash_to_outcomes(
```
Comment (bowenwang1996) [inline chain/chain/src/store.rs:228]: This name is confusing. What about `get_outcomes_by_block_hash`?

File: core/store/src/lib.rs (hunk 4 above)
Comment (bowenwang1996) [reply core/store/src/lib.rs:300]: We cannot due to the dependency here.

File: core/store/src/lib.rs (hunk 3 above)
Comment (bowenwang1996) [reply core/store/src/lib.rs:284]: Makes sense.

File: chain/chain/src/store.rs (hunk 1 above)
Comment (Kouprin) [reply chain/chain/src/store.rs:2445]: ++

File: chain/chain/src/store.rs (hunk 2 above)
Comment (Kouprin) [reply chain/chain/src/store.rs:3009]: why?

File: core/store/src/lib.rs (hunk 3 above)
Comment (Kouprin) [reply core/store/src/lib.rs:284]: ++

File: chain/chain/src/store.rs (hunk 5 above)
Comment (Kouprin) [reply chain/chain/src/store.rs:228]: ++

**Comment by Kouprin** [comment]
> @Kouprin there seems to be no test

What test you expect to be added?

**Comment by bowenwang1996** [comment]
@Kouprin can we test that execution outcomes are properly garbage collected?

**Comment by Kouprin** [comment]
> @Kouprin can we test that execution outcomes are properly garbage collected?

I think https://github.com/nearprotocol/nearcore/issues/2893 will cover all test cases related to this PR. Which checks are the most critical? I can add them to this PR.

File: chain/chain/src/store.rs (hunk 2 above)
Comment (frol) [reply chain/chain/src/store.rs:3007]: Why do we need this println?

Comment (Kouprin) [reply chain/chain/src/store.rs:3007]: It prints which column is processing now. If assert fails, it would be clear what's wrong even without local rerun.

File: chain/chain/src/store.rs [hunk 6]
```diff
@@ -1921,6 +1939,7 @@ impl<'a> ChainStoreUpdate<'a> {
     ) -> Result<(), Error> {
         let mut store_update = self.store().store_update();
         let header = self.get_block_header(&block_hash).expect("block header must exist").clone();
+        let next_epoch_id = header.epoch_id();
```
Comment (bowenwang1996) [inline chain/chain/src/store.rs:1942]: Is this intended (instead of `header.next_epoch_id()`)?

**Comment by SkidanovAlex** [comment]
The change seems to be rather invasive to at least have sanity tests (that in simple cases exactly the outcomes we expect are garbage collected)

File: chain/chain/src/store.rs [hunk 7]
```diff
@@ -1998,6 +2017,10 @@ impl<'a> ChainStoreUpdate<'a> {
             self.gc_col(ColNextBlockWithNewChunk, &height_shard_id);
             let key = StateHeaderKey(shard_id, block_hash).try_to_vec()?;
             self.gc_col(ColStateHeaders, &key);
+
+            if self.get_last_block_with_new_chunk(shard_id)? == Some(&block_hash) {
+                self.gc_col(ColLastBlockWithNewChunk, &index_to_bytes(shard_id));
+            }
```
Comment (SkidanovAlex) [inline chain/chain/src/store.rs:2023]: Are there any negative consequences of not GCing this column?
Let's only add GCing now for columns that can potentially result in memory consumption growing uncontrollably.

File: chain/chain/src/store.rs [hunk 8]
```diff
@@ -2037,6 +2066,14 @@ impl<'a> ChainStoreUpdate<'a> {
...
                 }
                 self.clear_chunk_data(min_chunk_height)?;
+                self.remove_state_dl_info(block_hash);
+                // Check Epoch switch
+                let cur_epoch_id = block.header().epoch_id();
+                if cur_epoch_id != next_epoch_id {
+                    // Current Block is the last one who needs current Epoch
+                    self.gc_col(ColEpochInfo, &cur_epoch_id.as_ref().into());
+                    self.gc_col(ColEpochStart, &cur_epoch_id.as_ref().into());
+                }
```
Comment (SkidanovAlex) [inline chain/chain/src/store.rs:2076]: Same here, let's defer this until after phase 1 (alternatively send this and the above as a separate PR against `post-phase-1` branch)

File: chain/chain/src/store.rs (hunk 7 above)
Comment (Kouprin) [reply chain/chain/src/store.rs:2023]: > Are there any negative consequences of not GCing this column?

Let's say it will not cleared and point to deleted Block after complete Sync. Did we consider this case while designing the column?

I guess it may be more useful to keep Storage in consistent state rather than have non-cleared columns specifically.

**Comment by Kouprin** [comment]
> The change seems to be rather invasive to at least have sanity tests (that in simple cases exactly the outcomes we expect are garbage collected)

I do believe it's much better to check Storage consistency than have a couple of sanity tests. Please have a look into https://github.com/nearprotocol/nearcore/issues/2893. It should cover most of reasonable cases.

File: chain/chain/src/store.rs (hunk 6 above)
Comment (Kouprin) [reply chain/chain/src/store.rs:1942]: Absolutely. On the Canonical chain we replace the current hash with previous one below.

File: chain/chain/src/store.rs (hunk 2 above)
Comment (frol) [reply chain/chain/src/store.rs:3007]: Then I would prefer having some accompanying message along side to the printed value then

Comment (Kouprin) [reply chain/chain/src/store.rs:3007]: Makes sense, ++.
BTW, I would like to point that printing some minor debug in tests is a good thing. I don't know why we used to avoid it.

Comment (frol) [reply chain/chain/src/store.rs:2995]: I believe, we should better invest into having enough debug tracing calls in the code, so we don’t need to have extra debug information in the tests. This way we will be able to debug issues on ruining instances, and not only in tests.

**Comment by Kouprin** [comment]
Nightly fails because of 

```

CMake Warning:
--
  | Manually-specified variables were not used by the project:
  |  
  | CMAKE_ASM_COMPILER
  | CMAKE_ASM_FLAGS
  |  
  |  
  | /home/bo/.cargo/registry/src/github.com-1ecc6299db9ec823/wabt-sys-0.7.2/wabt/src/option-parser.cc: In lambda function:
  | /home/bo/.cargo/registry/src/github.com-1ecc6299db9ec823/wabt-sys-0.7.2/wabt/src/option-parser.cc:60:20: error: ‘CMAKE_PROJECT_VERSION’ was not declared in this scope
  | printf("%s\n", CMAKE_PROJECT_VERSION);
  | ^~~~~~~~~~~~~~~~~~~~~
  | make[2]: *** [CMakeFiles/wabt.dir/src/option-parser.cc.o] Error 1
  | make[2]: *** Waiting for unfinished jobs....
  | make[1]: *** [CMakeFiles/wabt.dir/all] Error 2
  | make: *** [all] Error 2
  | thread 'main' panicked at '
  | command did not execute successfully, got: exit code: 2
  |  
  | build script failed, must exit now', /home/bo/.cargo/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.44/src/lib.rs:885:5
  | note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  |  
  | warning: build failed, waiting for other jobs to finish...
  | error: build failed


```

nearlib-test failed because of
```

[2020-06-25T07:59:56Z] FAIL test/promise.test.js
--
  | [2020-06-25T07:59:56Z]   ● Console
  | [2020-06-25T07:59:56Z]
  | [2020-06-25T07:59:56Z]     console.warn
  | [2020-06-25T07:59:56Z]       Failed to load master account key from /var/lib/buildkite-agent/.near/validator_key.json: Error: ENOENT: no such file or directory, open '/var/lib/buildkite-agent/.near/validator_key.json'

```

**Comment by frol** [comment]
@olonho is that CMake error known?

**Comment by birchmd** [comment]
The cmake bug been reported on the wabt-rs repo. Looks like we just need to update cmake to a later version on the buildkite runners.

https://github.com/pepyakin/wabt-rs/issues/69

File: chain/chain/src/store.rs (hunk 2 above)
Comment (Kouprin) [reply chain/chain/src/store.rs]: It's romantic. In fact, when I look into failing catching_up or gc test, only manual tracing creates most of understanding. 99% of standard debug is useless in those cases - not because it's useless overall, but because the test checks some specific things that need to be tested.

**Comment by ailisp** [comment]
> The cmake bug been reported on the wabt-rs repo. Looks like we just need to update cmake to a later version on the buildkite runners.

Yeah, you'll need to update to cmake 3.13 locally. I'll update in nightly runner

**Comment by ailisp** [comment]
Updated

**Comment by Kouprin** [comment]
@mfornet I've refactored State Sync types a little. Could you please review the commit https://github.com/nearprotocol/nearcore/pull/2891/commits/65b35ea7a75554affe0f8ecac72e5f1ff6b98bca ?
There will be some conflicts with https://github.com/nearprotocol/nearcore/pull/2515
Hope you merge first :)

**Comment by Kouprin** [comment]
The set of failing tests (https://build.nearprotocol.com.s3.amazonaws.com/nightly/output/T2860_204231547b3f0eef5a6b4fe201a6647d95719f87_20200626_065641/fails.txt) is the same as in master branch (http://nightly.neartest.com/run/master_fa93c794a81796e37db07cd425ef8b7c2633f619_20200625_160813)

Consider the change is valid.

File: chain/chain/src/store.rs [hunk 9]
```diff
@@ -2527,6 +2592,7 @@ impl<'a> ChainStoreUpdate<'a> {
             block_refcounts,
             block_merkle_tree,
             block_ordinal_to_hash,
+            gc_count: _,
```
Comment (bowenwang1996) [inline chain/chain/src/store.rs:2595]:
```suggestion
           ..
```

File: core/store/src/db.rs [hunk 10]
```diff
@@ -145,11 +147,46 @@ impl std::fmt::Display for DBCol {
...
+        col_gc[DBCol::ColEpochLightClientBlocks as usize] = false;
+        col_gc[DBCol::ColPeerComponent as usize] = false; // Peer related info doesn't GC
+        col_gc[DBCol::ColLastComponentNonce as usize] = false;
+        col_gc[DBCol::ColComponentEdges as usize] = false;
+        col_gc[DBCol::ColBlockOrdinal as usize] = false;
+        col_gc
+    };
+}
+
+// List of columns for which GC may not be executed even in fully operational node
```
Comment (bowenwang1996) [inline core/store/src/db.rs:177]: why are they not executed?

Comment (Kouprin) [reply core/store/src/db.rs:177]: I think about it in the following terms:
1. Column should implement GC and will use it.
2. Column should implement GC and it may not use it until specific circumstances.
3. Column should not be GCed.

Columns that are in SHOULD_COL_GC should implement their GC method anyway. However, it doesn't mean that it will be called in any predictable time in fully operational node. For example, with no State Sync called ColStateHeaders will be never cleared because they will not be created. That's why SKIP_COL_GC comes.

**Comment by Kouprin** [comment]
Ok, merging as it's blocking https://github.com/nearprotocol/nearcore/pull/2515 and gives us additional time to test plenty of Nightly Runs to check validity.
